### PR TITLE
Composite + Repeating As Core Scheming Features

### PR DESCRIPTION
**Pull request by TkTech**
A bit messy but as people were requesting the changes, here they are. Absolutely 100% not-ready-to-merge (no tests!) It's based on PR #176. It tries not to break too many templates which is why there are some real ugly kludges (ex: `scheming_massage_subfield`)

The main changes in this are:

1. Field default values, which can be simple text strings or complex jinja templates. Ex:
```yaml
- field_name: date_published
  label: Date Published
  default: "{{ h.now().strftime('%Y-%m-%d') }}"
  form_snippet: date.html
  validators: require_when_published
```
2. Support for grouping fields, both for display and editing (`display_group: General`) (#175)
3. Resources and packages are rendered identically, simplifying things.
4. The behaviour of ckanext-composite and ckanext-repeating is included in core, but is implemented completely differently. You can have complex, repeating fields and field groups using any snippets, autocompletes, validators, etc... and they'll all "just work". Just provide a list of `subfields` to turn a field into a composite field, and add `repeatable: true` to allow more than one. `form_blanks` can be used to specify how many repeatable fields should be rendered by default. Ex:
```yaml
- field_name: change_history
  label: Change History
  help_text: "A description of changes made to the data since its release. For each change, indicate the date of change and describe what was changed and why. When a dataset is first published, there is one change history entry for the initial creation of the dataset.  Add additional change history entries as needed."
  display_group: Science
  repeatable: true
  form_blanks: 1
  subfields:
  - field_name: change_description
    label: Description of Change
    default: "Initial creation of dataset record"
    form_snippet: textarea.html
    display_snippet: text.html
    validators: require_when_published

  - field_name: change_date
    label: Date of Change
    form_snippet: date.html
    display_snippet: date.html
    default: "{{ h.now().strftime('%Y-%m-%d') }}"
```
5. Missing call to the help text snippet is added to slug.html

@wardi requested you as a reviewer just to get your thoughts if you have time. I will likely squash this and redo it before anything gets merged so feedback is appreciated.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -20,7 +20,6 @@ dependencies:
   override:
   - git clone https://github.com/ckan/ckan
   - python setup.py develop
-  - pip install -r requirements.txt
```
Comment (wardi) [inline circle.yml]: why was this removed?

File: ckanext/scheming/commands.py [hunk 2]
```diff
@@ -1,65 +0,0 @@
...
-                field_names = ('dataset_fields', 'fields', 'resource_fields')
-
-                for field_name in field_names:
-                    if s[typ].get(field_name):
-                        if field_name == 'resource_fields':
-                            print " * " + json.dumps("resource")
-                        for field in s[typ][field_name]:
-                            print "   - " + json.dumps(field['field_name']),
-                            print scheming_language_text(field.get('label'))
-            print
```
Comment (wardi) [inline ckanext/scheming/commands.py]: removing to reduce paste dependencies?

File: circle.yml (hunk 1 above)
Comment (wardi) [reply circle.yml]: ah, moved to install_requires

**Comment by wardi** [comment]
moving the examples into their own directory is good, but our readme says to use  `scheming.dataset_schemas = ckanext.scheming:camel_photos.json` which will no longer work. Could move the examples dir to ckanext/scheming/examples instead so that this type of import will still work

File: ckanext/scheming/plugins.py [hunk 3]
```diff
@@ -69,65 +51,68 @@ class _SchemingMixin(object):
...
+                'scheming.presets',
+                DEFAULT_PRESETS
+            ).split()
+        )
+
+        _SchemingMixin._presets = {
+            field['preset_name']: field['values']
+            for preset_path in presets
+            for field in _load_schema(preset_path)['presets']
+        }
```
Comment (wardi) [inline ckanext/scheming/plugins.py]: I like renaming `f` to `preset_path` but otherwise I find the original code easier to understand

File: ckanext/scheming/plugins.py [hunk 4]
```diff
@@ -234,28 +220,44 @@ def validate(self, context, data_dict, schema, action):
...
-        if action_type == 'show':
-            get_validators = _field_output_validators
-        elif action_type == 'create':
-            get_validators = _field_create_validators
-        else:
-            get_validators = _field_validators
+        get_validators = {
+            'show': _field_output_validators,
+            'create': _field_create_validators,
+            'upate': _field_create_validators
```
Comment (wardi) [inline ckanext/scheming/plugins.py]: `upate`?  Also update doesn't use the `_field_create_validators`. I prefer the original code here even if it's longer.

File: ckanext/scheming/plugins.py [hunk 5]
```diff
@@ -234,28 +220,44 @@ def validate(self, context, data_dict, schema, action):
...
+                )
+
+                # Apply default field values before going through validation. This
+                # deals with fields that have form_snippet set to null, and fields
+                # that have defaults added after initial creation.
+                if data_dict.get(f['field_name']) is None:
+                    default = f.get('default')
+                    if default:
+                        data_dict[f['field_name']] = (
+                            helpers.scheming_render_from_string(
```
Comment (wardi) [inline ckanext/scheming/plugins.py]: I love jinja template defaults, but let's name them "default_template" instead of "default" so we still have the fast path for common non-jinja defaults.

File: ckanext/scheming/plugins.py [hunk 6]
```diff
@@ -362,10 +364,12 @@ def _load_schema_module_path(url):
...
-    p = os.path.join(os.path.dirname(inspect.getfile(m)), file_name)
-    if os.path.exists(p):
-        watch_file(p)
-        return loader.load(open(p))
+
+    path = os.path.join(os.path.dirname(inspect.getfile(m)), file_name)
+    if os.path.exists(path):
+        watch_file(path)
+        with open(path, 'rb') as schema_file:
+            return yaml.load(schema_file)
```
Comment (wardi) [inline ckanext/scheming/plugins.py]: I don't love parsing json with yaml because then users are going to sneak yamlisms into their schemas without renaming them. Also since pyyaml has code execution vulnerabilities this opens up sites with live schema editors to potential problems

File: ckanext/scheming/plugins.py [hunk 7]
```diff
@@ -412,16 +415,25 @@ def _field_validators(f, schema, convert_extras):
...
+            f['validators'],
+            f,
+            schema
+        )
     elif helpers.scheming_field_required(f):
-        validators = [not_empty, unicode]
+        validators = [not_empty]
     else:
-        validators = [ignore_missing, unicode]
+        validators = [ignore_missing]
```
Comment (wardi) [inline ckanext/scheming/plugins.py]: huh. I wonder why I had unicode there. Good change.

File: ckanext/scheming/helpers.py [hunk 8]
```diff
@@ -343,3 +347,46 @@ def scheming_display_json_value(value, indent=2):
...
+
+    return template.render(**kwargs)
+
+
+def scheming_massage_subfield(field, subfield, index, data):
+    # Subfield & data are re-used, we must not modify the original!
+    sf = subfield.copy()
+    data = data.copy()
+
+    new_field_name = '{field_name}-{index}-{subfield_name}'.format(
```
Comment (wardi) [inline ckanext/scheming/helpers.py]: I learned that our controllers already have a convention for nested parameters: 
https://github.com/ckan/ckan/blob/2.7/ckan/controllers/package.py#L564-L565

This converts parameters in the form `field_name__index__subfield_name` into lists of nested dicts automatically in the controller. We should use this instead of inventing another nested list-of-dicts implementation for scheming.

File: ckanext/scheming/templates/scheming/form_snippets/_organization_select.html [hunk 9]
```diff
@@ -10,7 +10,7 @@
     error=errors[field.field_name],
     is_required=org_required,
     classes=['form-group', 'control-medium'],
-    extra_html=caller() if caller,
```
Comment (wardi) [inline ckanext/scheming/templates/scheming/form_snippets/_organization_select.html]: trailing commas are my right as a python programmer gosh darn it! :-)

File: ckanext/scheming/templates/scheming/form_snippets/slug.html [hunk 10]
```diff
@@ -32,4 +32,4 @@
     is_required=h.scheming_field_required(field)
     ) }}
 
-
+{%- snippet 'scheming/form_snippets/help_text.html', field=field -%}
```
Comment (wardi) [inline ckanext/scheming/templates/scheming/form_snippets/slug.html]: was this getting repeated?

File: ckanext/scheming/templates/scheming/package/resource_read.html [hunk 11]
```diff
@@ -9,58 +9,61 @@
...
+                            object_type=dataset_type -%}
+                                </dd>
+                            {% endfor %}
+                        </dl>
+                    </div>
+                {% endif %}
+            {% endfor %}
+        {% endblock %}
+
+    </div>
```
Comment (wardi) [inline ckanext/scheming/templates/scheming/package/resource_read.html]: ckan's templates are indented 2 spaces, changing all the indentation to 4 is making it harder to see the actual change here

File: setup.py [hunk 12]
```diff
@@ -19,17 +15,17 @@
...
+        'pytz'
+    ],
     entry_points=\
     """
     [ckan.plugins]
     scheming_datasets=ckanext.scheming.plugins:SchemingDatasetsPlugin
     scheming_groups=ckanext.scheming.plugins:SchemingGroupsPlugin
     scheming_organizations=ckanext.scheming.plugins:SchemingOrganizationsPlugin
-    scheming_test_subclass=ckanext.scheming.tests.plugins:SchemingTestSubclass
-    scheming_test_plugin=ckanext.scheming.tests.plugins:SchemingTestSchemaPlugin
```
Comment (wardi) [inline setup.py]: removing test plugins?

File: ckanext/scheming/helpers.py (hunk 8 above)
Comment (wardi) [reply ckanext/scheming/helpers.py]: This could be too large a change to ask for as part of getting this PR merged. Let's just revisit later.

**Comment by wardi** [comment]
This PR just needs conflicts resolved and some minor fixes mentioned above.

File: ckanext/scheming/templates/scheming/form_snippets/slug.html (hunk 10 above)
Comment (TkTech) [reply ckanext/scheming/templates/scheming/form_snippets/slug.html]: It was added not removed, no help text was being displayed for slugs.

File: ckanext/scheming/helpers.py (hunk 8 above)
Comment (TkTech) [reply ckanext/scheming/helpers.py]: It's commented elsewhere, but it's done this way because that's already how composite and repeating do it. Without this, plugins that were built on top of them cannot find the fields.

File: ckanext/scheming/commands.py (hunk 2 above)
Comment (TkTech) [reply ckanext/scheming/commands.py]: Needs paste and an update for py3. Didn't seem terribly useful so with limited time I just removed it.

File: ckanext/scheming/plugins.py (hunk 5 above)
Comment (TkTech) [reply ckanext/scheming/plugins.py]: Which takes precedence when overloading from presets?

File: ckanext/scheming/plugins.py (hunk 3 above)
Comment (TkTech) [reply ckanext/scheming/plugins.py]: There was a reason for this, to make the assignment a single atomic operation...can't remember why tho, I think something I was experimenting with in load_schema would call something else that would in turn look at the presets again...

File: ckanext/scheming/plugins.py (hunk 5 above)
Comment (wardi) [reply ckanext/scheming/plugins.py]: let's say default_template since it's cooler. If there's a default_template in a preset but a schema wants a static value they can either use the same static value as the new default_template or set `default_template: null`

File: ckanext/scheming/plugins.py (hunk 3 above)
Comment (wardi) [reply ckanext/scheming/plugins.py]: ok, so we could build this as a new dict then overwrite the shared one to make it atomic.

The way these values are cached in class variables probably isn't safe even with the atomic update for example with tests reloading configurations. Need a better way to store and update these.

File: ckanext/scheming/commands.py (hunk 2 above)
Comment (wardi) [reply ckanext/scheming/commands.py]: I'm fine with this

**Comment by wardi** [comment]
closing in favour of #203

File: ckanext/scheming/templates/scheming/package/resource_read.html [hunk 13]
```diff
@@ -9,58 +9,61 @@
...
+                    <th scope="row">{{ _('License') }}</th>
+                    <td>{% snippet "snippets/license.html", pkg_dict=pkg, text_only=True %}</td>
+                </tr>
+            {%- endblock -%}
+            </tbody>
+        </table>
+
+        {% block resource_fields %}
+            {%- for group in schema.resource_fields|groupby('display_group') -%}
+                {%- set group_fields = h.non_empty_fields(group.list|list, res, exclude_fields) -%}
```
Comment (wardi) [inline ckanext/scheming/templates/scheming/package/resource_read.html]: Where's this defined?

Comment (TkTech) [reply ckanext/scheming/templates/scheming/package/resource_read.html]:
```
def non_empty_fields(field_list, pkg_dict, exclude):
    r = []
    for field in field_list:
        if field['field_name'] in exclude:
            continue

        if field.get('display_snippet', False) is None:
            continue

        if pkg_dict.get(field['field_name']):
            r.append(field)
return r
```

Was missed when pulling chunks out for the PR.
